docs/design/sparse-v2.md

```diff
@@ -0,0 +1,225 @@
+
+```rust
+pub enum SparsePatternsPathType {
+    DirInclusion,  // Includes <path>/...
+    DirExclusion,  // Excludes <path>/...
+    Files,         // Includes <path>/*
+    File,          // Includes <path> exactly
+}
+
+pub struct SparsePatternsPath {
+    type: SparsePatternsPathType,
+    path: RepoPathBuf,
+}
+
+pub struct SparsePatternsMapping {
+    recursive: bool,  // If false, only immediate children of src_path (files) are renamed.
+    src_path: RepoPathBuf,
+    dst_path: RepoPathBuf,
+}
+
+pub struct SparsePatterns {
+    paths: Vec<SparsePatternsPath>,
+    mappings: Vec<SparsePatternsMapping>,
+}
+
+pub trait OpStore {
+    ...
+    pub fn read_sparse_patterns(&self, id: &SparsePatternsIn) -> OpStoreResult<SparsePatterns> { ... }
+    pub fn write_sparse_patterns(&self, sparse_patterns: &SparsePatterns) -> OpStoreResult<SparsePatternsId> { .. }
+}
+```
+
+To support these more complex behaviors, a new `SparsePatterns` trait will be
+introduced, mainly to facilitate migration to the new data type at all the
+places where sparse patterns are currently used.
+
+```rust
+impl SparsePatterns {
+    pub fn to_matcher(&self) -> Box<dyn Matcher> {
+        ...
+    }
+
+    ...
+}
+```
+
+## Command Syntax
+
+Path rules can be specified on the CLI and in an editor via a `<type>:<path>`
+syntax. The editor will only support the new explicit syntax, whereas the CLI
+arguments will continue to support bare `<path>` arguments in a
+backwards-compatible way at least for a migration period.
+
+```rust
+pub fn type_string(type: SparsePatternsPathType) -> &'static str {
+    match type {
+        SparsePatternsPathType::DirInclusion => "include",
+        SparsePatternsPathType::DirExclusion => "exclude",
+        SparsePatternsPathType::Files => "files",
+        SparsePatternsPathType::File => "file",
+    }
+}
+```
+
+- `jj sparse set --add foo/bar` is equal to `jj sparse set --add include:foo/bar`
+- `jj sparse set --add exclude:foo/bar` adds a new `DirExclusion` type rule
+- `jj sparse set --exclude foo/bar` as a possible shorthand for the above
+- `jj sparse list` will print the explicit rules
+
+Paths will be stored in an ordered, canonical form which unambiguously describes
+the set of files to be included. Every `--add` command will append to the end of
+this list before the patterns are canonicalized. Whether a file is included is
+determined by the first matching rule in reverse order.
+
+For example:
+
+```txt
+include:foo
+exclude:foo/bar
+include:foo/bar/baz
+exclude:foo/bar/baz/qux
+```
+
+Produces rule set which includes "foo/file.txt", excludes "foo/bar/file.txt",
+includes "foo/bar/baz/file.txt", and excludes "foo/bar/baz/qux/file.txt". If the
+rules are subtly re-ordered, they become canonicalized to a smaller, but
+functionally equivalent form:
+
+```txt
+# Before
+include:foo
+exclude:foo/bar/baz/qux
+include:foo/bar/baz
+exclude:foo/bar
+
+# Canonicalized
+include:foo
+exclude:foo/bar
+```
```

Comment:
Thanks for expanding on the canonicalization. Instead of repeating my question again, I just want to ask if any other strategies were considered and discarded. If there were any it be worth a mention, but I'm consider my concerns adressed.

Reply:
I'm open to other ideas; I didn't really consider any other than what was in previous iterations of this doc. The main constraints I have in mind are:

- Needs to have backwards-compatible syntax
- Needs to support this type of expressive include+exclude nesting (for internal integration)
- Should be simple and easily comprehensible otherwise

I briefly considered some sort of structured tree-stanza syntax closer to what an implementation would use to actually evaluate paths, but that's a huge pain for an end-user to construct/edit.  Did you have another idea in mind?

Reply:
> I'm open to other ideas; I didn't really consider any other than what was in previous iterations of this doc. The main constraints I have in mind are:
> 
> * Needs to have backwards-compatible syntax
> * Needs to support this type of expressive include+exclude nesting (for internal integration)
> * Should be simple and easily comprehensible otherwise
> 
> I briefly considered some sort of structured tree-stanza syntax closer to what an implementation would use to actually evaluate paths, but that's a huge pain for an end-user to construct/edit.

Could you add a summary of the above to the doc as the reason why we choose or only considered this design? That is basically everything I wanted from my initial comment on. 

> Did you have another idea in mind?

I'm all out of ideas at the moment which could fit into this design with its constraints, so I won't block on it. Perhaps we notice something which we can improve during the implementation.